TAN-2276 Nav bar item as project

IvaKop opened the pull request:
# Changelog
<!-- Replace this comment by a bullet list. More info: https://www.notion.so/citizenlab/Changelog-How-it-works-f418426c75994454a332bf067634f3f1 -->

IvaKop committed:
Add migration

IvaKop committed:
Do migration

IvaKop committed:
Fix request

IvaKop committed:
Add be changes

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
Fix links

IvaKop committed:
Merge branch 'TAN-2276/nav-bar-item-as-project' of github.com:CitizenLabDotCo/citizenlab into TAN-2276/nav-bar-item-as-project

IvaKop committed:
Fix test

IvaKop committed:
Fix form

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
Filter based on visibility

IvaKop committed:
Merge branch 'TAN-2276/nav-bar-item-as-project' of github.com:CitizenLabDotCo/citizenlab into TAN-2276/nav-bar-item-as-project

IvaKop committed:
Final touches

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
Fix tests

IvaKop committed:
Merge branch 'TAN-2276/nav-bar-item-as-project' of github.com:CitizenLabDotCo/citizenlab into TAN-2276/nav-bar-item-as-project

IvaKop committed:
Fix back lint

IvaKop committed:
Fix last test

IvaKop committed:
Fix visible navbar items

IvaKop committed:
Add tooltip

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
Fix number

IvaKop committed:
Merge branch 'TAN-2276/nav-bar-item-as-project' of github.com:CitizenLabDotCo/citizenlab into TAN-2276/nav-bar-item-as-project

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
Fix unit test

IvaKop committed:
Merge branch 'TAN-2276/nav-bar-item-as-project' of github.com:CitizenLabDotCo/citizenlab into TAN-2276/nav-bar-item-as-project

IvaKop committed:
Fix double active state for pinned projects

sebastienhoorens commented on back/app/models/nav_bar_item.rb:
```diff
@@ -63,12 +76,15 @@ def set_code
 
   def fallback_title_multiloc
-    key_code = custom? ? static_page.code : code
+    page? ? static_page.code : code
```
Could it be something is wrong with this line? Should we also keep using the key_code for pages?

IvaKop replied:
Yes, forgot to assign the `key_code` to this. Thanks for catching it!

sebastienhoorens commented on back/app/serializers/web_api/v1/nav_bar_item_serializer.rb:
```diff
@@ -8,4 +8,5 @@ class WebApi::V1::NavBarItemSerializer < WebApi::V1::BaseSerializer
   end
 
   belongs_to :static_page, serializer: WebApi::V1::StaticPageSerializer
+  belongs_to :project, serializer: ::WebApi::V1::ProjectSerializer
```
Do you think the frontend needs to make a separate request to get the project (or static page)? If so, we might want to consider including these relationships in the `GET nav_bar_items` request.

IvaKop replied:
You mean as included resources? I don't think there's need for this (at least for now).

sebastienhoorens commented on front/app/containers/Admin/pagesAndMenu/containers/AddProjectNavbarItemModal/index.tsx:
```diff
@@ -0,0 +1,139 @@
+};
+
+const AddProjectNavbarItemModal = ({ opened, onClose }: Props) => {
+  const { mutateAsync: addNavbarItem } = useAddNavbarItem();
```
I don't fully understand what this line does. It looks like a destructor pattern and assignment, so I would expect `mutateAsync` or `addNavbarItem` to be new variable names, but `addNavbarItem` is the name of a function that is already defined.

IvaKop replied:
It just means that the `mutateAsync` method that comes from the hook `useAddNavbarItem` gets renamed to `addNavbarItem` for clarity and more readability down the line when it's used. It's always a good idea to name these methods properly, especially when there's more than one hook being used (not the case here).

sebastienhoorens commented on front/app/containers/Admin/pagesAndMenu/containers/AddProjectNavbarItemModal/index.tsx:
```diff
@@ -0,0 +1,139 @@
+  });
+
+  const methods = useForm({
+    mode: 'onBlur',
```
What does`mode: 'onBlur'` do? (it seems to be related to the form losing focus, but I can't figure out what this could mean in this context)

IvaKop replied:
This defined the validation method for the form fields. It means the validation rules will trigger when the field is blurred. Another way could be `onChange` for example but we use blur because it's better UX.

sebastienhoorens commented on front/app/containers/Admin/pagesAndMenu/containers/AddProjectNavbarItemModal/index.tsx:
```diff
@@ -0,0 +1,139 @@
+      await addNavbarItem({ ...formValues, type: 'project' });
+      handleOnClose();
+    } catch (error) {
+      handleHookFormSubmissionError(error, methods.setError);
```
I don't fully understand what this line does either 😅 Does it convert a raised Javascript exception to a readable error message?

IvaKop replied:
This is a method that transforms the api error message into a human readable message provided we have this exception in our list of errors for which we have copy. If not - we show a generic error message.

sebastienhoorens commented on front/app/containers/Admin/pagesAndMenu/containers/AddProjectNavbarItemModal/index.tsx:
```diff
@@ -0,0 +1,139 @@
+            <Box display="flex" gap="32px" flexDirection="column">
+              <Select
+                name="projectId"
+                label="Project"
```
Shouldn't we use `formatMessage` here?

IvaKop replied:
Absolutely! 👍

sebastienhoorens commented on front/app/containers/Admin/pagesAndMenu/containers/NavigationSettings/HiddenNavbarItemList/index.tsx:
```diff
@@ -101,11 +101,7 @@ const HiddenNavbarItemList = ({
-                  ? item.navbarTitleMultiloc
-                  : item.pageTitleMultiloc
-              }
+              title={item.titleMultiloc}
```
Do you think this could cause the names of page items on existing platforms to change (which would probably not be desired)?

IvaKop replied:
I don't think it will cause an issue. It's just a streamlining the properties of the items a bit to simplify.

sebastienhoorens commented on front/app/containers/Admin/pagesAndMenu/containers/PagesMenu/index.tsx:
```diff
@@ -1,36 +1,69 @@
+    return null;
+  }
+
+  const disabledAddProjectToNavbarButton = navbarItems.data.length >= 7;
```
Nit: I wonder if it could be valuable to have a helper method (for example a predicate) to determine if the number of items were exceeded, so it can be reused in other places, like knowing if we can add a static page.

IvaKop replied:
Good point! I don't think it's worth it's own utility but I added a constant for the max number of nav items to make it clear where this number comes from.

sebastienhoorens commented on front/app/containers/Admin/pagesAndMenu/containers/PagesMenu/messages.ts:
```diff
@@ -14,4 +14,12 @@ export default defineMessages({
+  },
+  navBarMaxItems: {
+    id: 'app.modules.navbar.admin.containers.navBarMaxItemsNumber',
+    defaultMessage: 'You can only add up to 5 items to the navigation bar',
```
Nit: I wonder if "5 items" could be confusing if the admin sees 7 items (including home and all projects).

IvaKop replied:
I had it at 7 initially then saw the copy from Rob said 5 so I changed it. I guess let's go with this and see if it causes confusion. It's just copy so we can easily adjust it.

sebastienhoorens commented on front/app/containers/MainHeader/Components/DesktopNavItems/AdminPublicationsNavbarItem.tsx:
```diff
@@ -159,6 +159,12 @@ const AdminPublicationsNavbarItem = ({
 
   const urlSegments = location.pathname.replace(/^\/+/g, '').split('/');
   const secondUrlSegment = urlSegments[1];
+  const thirdUrlSegment = urlSegments[2];
```
I'm having trouble understanding what these `urlSegments` represent. Is this playing a role in hiding the project items if they are not visible?

IvaKop replied:
Basically, it is checking if there is something after `/projects/` and in this way it determines whether the `All projects` navbar item is active or not. I introduced this mechanism because if a project is pinned both the project item and `All projects` item in the navigation bar were active and this looked confusing.

sebastienhoorens commented on front/app/containers/MainHeader/Components/DesktopNavItems/index.tsx:
```diff
@@ -67,15 +70,17 @@ const DesktopNavItems = () => {
-              key={i}
-            />
-          );
+          if (linkTo) {
```
When can `linkTo` be `null`?

IvaKop replied:
When the visibility rules kick in and the user does not actually have access to this project. Then the link is `null` and we don't render the nav bar item.

sebastienhoorens reviewed (commented):
I think there might be an issue with `fallback_title_multiloc` (although I cannot reproduce a corresponding issue), so not approving yet, but it's probably good enough to release. It's very nice to see the frontend code. I think in general everything is cleanly implemented. I have a fair amount of questions around understanding the frontend, sorry 😅

Functional review:
- In general it feels really smooth, I really like it.
- The "Delete" button didn't work for me for projects (after reordering the project item).
- When editing, you cannot go back through the UI.
- When reordering, it sometimes happens that the effect is moving to the bottom, while the item is actually dragged to a different place.
- When creating a new project item, I only need to specify one locale, but when editing I need to specify all locales.

IvaKop committed:
Address PR comments

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
Fix validation

IvaKop commented:
@sebastienhoorens Thanks for the review, I addressed your comments. 
> The "Delete" button didn't work for me for projects (after reordering the project item).

Tried to re-order the projects and delete but couldn't reproduce the issue you encountered.

> When editing, you cannot go back through the UI

We have breadcrumbs for this so you can go back by clicking them.

> When reordering, it sometimes happens that the effect is moving to the bottom, while the item is actually dragged to a different place.

Couldn't really reproduce this. Do you have more details about when this happens?

> When creating a new project item, I only need to specify one locale, but when editing I need to specify all locales.

Good point, I changed the validation rules to be consistent 👍

Commit:
Translations updated by CI (extract-intl)

IvaKop committed:
Fix lint

IvaKop committed:
Merge branch 'TAN-2276/nav-bar-item-as-project' of github.com:CitizenLabDotCo/citizenlab into TAN-2276/nav-bar-item-as-project

sebastienhoorens commented on back/app/models/nav_bar_item.rb:
```diff
@@ -63,12 +76,14 @@ def set_code
 
   def fallback_title_multiloc
-    key_code = custom? ? static_page.code : code
+    key_code = page? ? static_page.code : project? ? nil : code
```
Ah so for projects there is no translation to look up write? In that case perhaps we should avoid trying `I18n.exists? key` as we already know it will return false. But not blocking.

sebastienhoorens reviewed (approved):
Good to go. I just have a small remark where it might be desirable to avoid calling `I18n.exists? key` for projects (I don't know to what extend this could impact performance), but I guess it will do string comparisons in a long list of translation keys and go through the whole list to conclude that there are no matches.

Thanks a lot for answering my beginner questions on the frontend!

I can show you my issues that I mentioned for the functional review if you like.

sebastienhoorens committed:
fallback title when project and remove delete button for project items
